### PR TITLE
Colorizer

### PR DESCRIPTION
**Pull request by joshmoore**
For supporting the application of lookup tables other than those provided in datasets, a new ReaderWrapper `Colorizer` has been added which takes an implementation of the new interface `LutSource` to convert one `byte[]` to another. The "standard" LUTs provided by ImageJ1's `LutLoader` are supported.

Testing:
 * `./bfconvert -colorize fire INPUT OUTPUT` should produce results similar to applying a LUT in ImageJ
 * `./showinf -colorize ice INPUT` show show an image with the alternative coloring.

Note: test OUTPUT should initially be of type `.ome.tiff`. `.png` is not yet working.

### CODE REVIEW

**Comment by melissalinkert** [comment]
From a conceptual/API perspective, my only real worry is that ```applyLUT``` in LutSource does not have enough information about what the input byte array represents.  ```x``` and ```y``` are probably unnecessary, but it seems like the bit depth and endianness should be passed through.    I would sort of assume that ```applyLUT``` is only operating on a single channel (though it's not enforced) - if the intention is for it to work across multiple channels, then the number and configuration of the channels would also need to be passed through.

From the model/formats meeting today, the corresponding Trello card is https://trello.com/c/Hv80rRdO/127-lookup-tables and I will work on opening one or more cleanup PRs against this next week.

**Comment by joshmoore** [comment]
Excluding to let #1555 through.

File: components/formats-api/src/loci/formats/BytesWrapper.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  {
+    int ch = getRGBChannelCount();
+    int bpp = FormatTools.getBytesPerPixel(getPixelType());
+    byte[] newBuffer;
+    try {
+      newBuffer = DataTools.allocate(w, h, ch, bpp);
+    }
+    catch (IllegalArgumentException e) {
+      throw new FormatException("Image plane too large. Only 2GB of data can " +
+        "be extracted at one time. You can workaround the problem by opening " +
```
Comment (mtbc) [inline components/formats-api/src/loci/formats/BytesWrapper.java]: "work around" as a verb

File: components/formats-api/src/loci/formats/BytesWrapper.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        "the plane in tiles; for further details, see: " +
+        "http://www.openmicroscopy.org/site/support/faq/bio-formats/" +
+        "i-see-an-outofmemory-or-negativearraysize-error-message-when-" +
+        "attempting-to-open-an-svs-or-jpeg-2000-file.-what-does-this-mean", e);
```
Comment (mtbc) [inline components/formats-api/src/loci/formats/BytesWrapper.java]: We need shorter versions of these URLs. :smiley:

**Comment by pwalczysko** [comment]
Note: this caused all-imports-on-trout-failure today. Maybe breaking ?

File: components/formats-api/src/loci/formats/BytesWrapper.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import java.io.IOException;
+
+import loci.common.DataTools;
+
+/**
+ * A {@link ReaderWrapper} helper which intercepts all
+ * openBytes-like methods to ease subclassing. The primary
+ * difficulty is that internally the various methods call one
+ * another. If one method is not intercepted, then the raw
+ * implementation might be accessed. The implementations here
```
Comment (mtbc) [inline components/formats-api/src/loci/formats/BytesWrapper.java]: This does seem like a significant risk.

File: components/formats-common/src/loci/common/lut/ij/LUT.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
+// Public domain
+// Copied from https://github.com/imagej/imagej1/blob/26fb68109dda7b429526d93bf2368ce28b8b07f9/ij/process/LUT.java
+// with thanks to Wayne Rasband
+
+package loci.common.lut.ij;
+
+import java.awt.Color;
+import java.awt.image.IndexColorModel;
+import java.text.DecimalFormat;
```
Comment (mtbc) [inline components/formats-common/src/loci/common/lut/ij/LUT.java]: With luck our French-locale CI testing will keep an eye on this.

File: components/formats-api/src/loci/formats/BytesWrapper.java (hunk 2 above)
Comment (hflynn) [inline components/formats-api/src/loci/formats/BytesWrapper.java]: Crap, I didn't release we had links to the FAQs in the code as well :(

Comment (mtbc) [inline components/formats-api/src/loci/formats/BytesWrapper.java]: Also one in BF `FormatReader`.

**Comment by melissalinkert** [comment]
PR opened against this to make the initial test case work: https://github.com/joshmoore/bioformats/pull/8

origin/develop will also need to be merged in before this can be included again.

**Comment by joshmoore** [comment]
Last few commits, @melissalinkert, start approaching a working `./showinf` but still getting odd effects. I assume we need more methods implemented a la https://github.com/joshmoore/bioformats/commit/aeac46149926577b2cf542199a7bf8ecab41f82f

**Comment by joshmoore** [comment]
I've pushed a building, more-or-less working version. Note the comment from https://github.com/joshmoore/bioformats/commit/ff4f840e052e58a236952548fc999850a31ca4e6 that it wasn't possible to pass a `CoreMetadata`.

At this point, bfconvert and showinf are testable and even sometimes working. OMERO is still showing no change in output. I'll unexclude the OMERO PR so that it's possible to test the whole stack.

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OME-5.1-merge-push#771](http://ci.openmicroscopy.org/job/OME-5.1-merge-push/771/). See the [console output](http://ci.openmicroscopy.org/job/OME-5.1-merge-push/771/consoleText) for more details.
Possible conflicts:
  - PR #1522 melissalinkert 'NFS performance'
    - components/formats-api/src/loci/formats/FormatReader.java
  - Upstream changes
    - components/formats-api/src/loci/formats/FormatReader.java

**Comment by joshmoore** [comment]
Fixed the conflict due to the FAQ URL changes and pushed.

File: components/formats-common/src/loci/common/lut/ij/LUT.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	public synchronized Object clone() {
+		try {return super.clone();}
+		catch (CloneNotSupportedException e) {return null;}
+	}
+
+	public  String toString() {
+		return "rgb[0]="+colorToString(new Color(getRGB(0)))+", rgb[255]="
+			+colorToString(new Color(getRGB(255)))+", min="+String.format("%.4f", min)+
+      ", max="+String.format("%.4", max);
```
Comment (melissalinkert) [inline components/formats-common/src/loci/common/lut/ij/LUT.java]: Findbugs complains on this line: https://ci.openmicroscopy.org/view/Bio-Formats/job/BIOFORMATS-5.1-merge-build/701/findbugsResult/new/?

**Comment by joshmoore** [comment]
Re-adding `exclude` to let gh-1621 through and work on int16.

**Comment by joshmoore** [comment]
Closing until we have more time for this.

**Pull request closed without merging**